### ISSUE TITLE
footmisc 包的 perpage 选项会在最后一行的末尾添加两个空格

### ISSUE THREAD

**Issue by tanukihee** [CTeX-org/forum#144]
## 检查

- [x] 已在 issues 中进行搜索（包括已关闭的问题）

## 编译环境

- 操作系统
  - [x] Windows 10
  - [ ] Windows 8/8.1
  - [ ] Windows 7
  - [ ] 更早版本的 Windows
  - [ ] macOS
  - [ ] Linux（请附发行版）

- TeX 发行版
  - [x] TeX Live 2020
  - [ ] MiKTeX 
  - [ ] CTeX 套装 2.9.2.164
  - [ ] 更早版本的 CTeX 套装

## 描述问题

如题

## 最小工作示例（MWE）

```latex
\documentclass{article}
\usepackage[showframe]{geometry}
\usepackage[perpage]{footmisc}
\begin{document}
\hfill lorem

\hfill ipsum
\end{document}
```

（编译后）输出

![image](https://user-images.githubusercontent.com/34062491/91564537-58e2f080-e973-11ea-8ce1-700276a818b9.png)

xe/lua/pdf 均能复现。

## 我还做了这些尝试

+ 修改 `\spaceskip` 的长度，末尾的空格长度亦会改变，计算出为两个空格的宽度。

+ 在最后一段后加 `%`，空格不消失。

+ 使用 footmisc 包其他选项，空格消失。

+ 在最后一段后、 `\end{document}` 前加一空行，空格消失。

**Comment by muzimuzhi** [comment]
是 `footmisc` 宏包的问题，看起来存在了多年，可以反馈给维护者。

1. 第一个多余空格，来自  `\AtEndDocument` 中未注释的换行符。可以在 `footmisc` 加载后，patch `\@enddocumenthook` 替换掉空格。
```tex
% footmisc.sty 2011/06/06 v5.5b, lines 547--551
  \AtEndDocument{\let\FN@pp@footnote@aux\FN@pp@footnote@endaux
    \def\FN@pp@lastfoot{-1}%
    \FN@pp@footnotehintfalse
    \renewcommand{\FN@pp@prev@foot}{root}
  } 
```

2. 第二个多余空格，看起来是 `footmisc` 修改后的 `\clearpage` 把末段末行的空格留在了 hmode，没有及时进入 vmode。这里我不够确定。瞎试了试，（在 `\footnotehint` 的定义里）用 `\immediate\write` 替代 `\protected@write` 能避免这个问题。

相关源码（`\end{document}` 会执行一次 `\clearpage`）
```tex
\let\FN@pp@@clearpage\clearpage
\renewcommand{\clearpage}{\footnotehint\FN@pp@@clearpage}

\newcommand{\footnotehint}{%
  \setcounter{footnote}{0}%
  % \immediate\write\@auxout{\noexpand\FN@pp@footnotehinttrue}%
  \@bsphack
  \protected@writeaux\relax{\protect\FN@pp@footnotehinttrue}%
  \@esphack
  \@tempcnta\c@@fnserial
  \advance\@tempcnta\@ne
  \global\c@pp@next@reset\@tempcnta
}
```

一个尝试的解决方案
```tex
\documentclass{article}
\usepackage[showframe]{geometry}
\usepackage[perpage]{footmisc}

\usepackage{xpatch}
\makeatletter
\xpatchcmd\@enddocumenthook
  {\renewcommand{\FN@pp@prev@foot}{root} }
  {\renewcommand{\FN@pp@prev@foot}{root}}
  {}{\fail}

\xpatchcmd\footnotehint
  {\protected@writeaux\relax{\protect\FN@pp@footnotehinttrue}}
  {\immediate\write\@auxout{\noexpand\FN@pp@footnotehinttrue}}
  {}{\fail}
\makeatother

\begin{document}
a \hfill lorem

a\hfill lipsum
\end{document}
```

**Comment by tanukihee** [comment]
剑桥邮箱拒收了😂，怕是很难联系到作者了

**Comment by zepinglee** [comment]
可能得用 .edu 后缀的邮箱

**Comment by tanukihee** [comment]
> 可能得用 .edu 后缀的邮箱

也退回来了😂

**Comment by zepinglee** [comment]
https://ctan.org/pkg/footmisc 这里显示 Robin Fairbairns 已经是 inactive 了，应该提到 https://github.com/FrankMittelbach/fmitex/issues

**Comment by tanukihee** [comment]
感谢，报告了

**Comment by FrankMittelbach** [comment]
> ```
> \xpatchcmd\footnotehint
>   {\protected@writeaux\relax{\protect\FN@pp@footnotehinttrue}}
>   {\immediate\write\@auxout{\noexpand\FN@pp@footnotehinttrue}}
>   {}{\fail}
> \makeatother
> ```

can one of you give me an explanation why you think this is needed? thanks

**Comment by muzimuzhi** [comment]
@FrankMittelbach 
I don't know why it works. Guess the `\protected@writeaux` makes the end-of-line space stayed in hmode.

**Comment by FrankMittelbach** [comment]
it would help if I would have an example where it fails. The one I got works correctly with pdftex, xetex and luatex

**Comment by muzimuzhi** [comment]
> it would help if I would have an example where it fails. The one I got works correctly with pdftex, xetex and luatex

What can be observed is, the `\immediate` changes the order of lines written to `.aux`. When the current page contains `\footnote`, before the patch `.aux` contains
```tex
\relax
\providecommand {\FN@pp@footnotehinttrue }{}
\providecommand {\FN@pp@footnote@aux }[2]{}
\FN@pp@footnote@aux{1}{1}
\FN@pp@footnotehinttrue
```
with `\immediate` applied `.aux` becomes
```tex
\relax
\FN@pp@footnotehinttrue  % <<< see here
\providecommand {\FN@pp@footnotehinttrue }{}
\providecommand {\FN@pp@footnote@aux }[2]{}
\FN@pp@footnote@aux{1}{1}
```
Note that `\FN@pp@footnote@aux` uses `\ifFN@pp@footnotehint` so I guess `\immediate` _will_ break something. I didn't read through the `footmist.sty`, hence I cannot provide a specific example yet. 

PS: Does it hurt if `\protected@writeaux\relax{\protect\FN@pp@footnotehinttrue}` is appended, not prefixed to `\clearpage`?

**Comment by FrankMittelbach** [comment]
that clearly means the patch is not going to work. The intention of the \providecommand being to establish the two commands if they are not there so they have to be first. Whether or not that make much sense is a different matter :-(.
But my point was, does anybody have an example where the issue shows up (preferably using std LaTeX not something specfic to ctex?) because right now what I got works fine with std LaTeX. I understand why the missing `%` is a problem but I fail to see why you need the second patch.

Basically I inherited  that package and at some point want to alter it drastically (using the new hook management etc) but not digging into its 30 years of history to be honest. Thus to make a change other than adding the percent I really would hope for an example to work with.

**Comment by muzimuzhi** [comment]
> But my point was, does anybody have an example where the issue shows up (preferably using std LaTeX not something specfic to ctex?)

The initially attached example in https://github.com/CTeX-org/forum/issues/144#issue-688072127 already has nothing specific to ctex and this problem is reproducible with overleaf's texlive 2019.

**Comment by FrankMittelbach** [comment]
oh well, I now se that what I changed (adding \showoutput and a blank line) that made the problem go away :-( ... ok got it, thanks

**Comment by muzimuzhi** [comment]
Also, the second extra space is reproducible with a whatsit inserted by `\label`, which I guess is not a new problem (or even documented, though I didn't find it yet).
```tex
\documentclass{article}
\usepackage[pass, showframe]{geometry}

\begin{document}
begin \hfill end \label{key}
\end{document}
```

**Comment by FrankMittelbach** [comment]
Yes this is the real cause. If `\end{document}`is called directly after the text (as in the example) it is still executed in hmode and the first thing it does is running the `\AtEndDocument` hook, so if that contains anything that surives (like a *non*-immediate `\write`that will become part of the paragraph and so the space between the last word and the `\end` will too. Your change to use `\immediate\write` makes the write go away, but the issue is general. So either, the answer is *add a blank line* in such case, or we add `\par`at the beginning of `\enddocument`to force it to be always into vmode. But there may be documents that relied on it being in hmode --- so not sure yet.

**Comment by muzimuzhi** [comment]
Related: https://github.com/latex3/latex2e/issues/385

**Comment by muzimuzhi** [comment]
沿用 https://github.com/CTeX-org/forum/issues/144#issuecomment-682839822 里的说法，
 - 第二个空格已经由 latex3/latex2e#385 修复，发布在 latex2e [dev-2021-05-01-1](https://github.com/latex3/latex2e/releases/tag/dev-2021-05-01-1)
 - 但因为 latex2e 2020-10-01 引入的 hook management，对第一个空格的 patch 已经失效。
目前，在 `footmisc.sty` 里使用 `\AtEndDocument{...}`，最终会在 `\g__hook_enddocument_code_prop` 末尾全局添加 `\__prop_pair:wn footmisc\s__prop{...}`，其中 `footmisc` 部分每个字符的 catcode 是 12（other），而不是一般的 11（letter），相当于使用了 `\detokenize{footmisc}。这导致既有的 patch 功能的宏包都用不了。

**Comment by FrankMittelbach** [comment]
>     * 但因为 latex2e 2020-10-01 引入的 hook management，对第一个空格的 patch 已经失效。
>       目前，在 `footmisc.sty` 里使用 `\AtEndDocument{...}`，最终会在 `\g__hook_enddocument_code_prop` 末尾全局添加 `\__prop_pair:wn footmisc\s__prop{...}`，其中 `footmisc` 部分每个字符的 catcode 是 12（other），而不是一般的 11（letter），相当于使用了 `\detokenize{footmisc}。这导致既有的 patch 功能的宏包都用不了。

I tried to understand that using google translate but unfortuately the translation becomes incomprehensible. If there is anything that is not covered by the update in latex-dev I would need an english explanation (and a test file) please.
It it is just about an internal patch that would now need updating, please ignore.

**Comment by muzimuzhi** [comment]
@FrankMittelbach Sorry for the inconvenience. I made a mistake by saying "the first space is not eliminated". The fact is, the problem reported here has been (totally) resolved in latest latex2e-dev released on 2021-01-11.

The quoted lines in your previous comments are about how to patch staffs added to `\AtEndDocument`. Prior to latex2e 2020-10-01, `\AtEndDocument` just appends its argument to `\@enddocumenthook`, which is rather easy to patch. Since 2020-10-01, `\AtEndDocument` used in `.sty` files will append codes to `\g__hook_enddocument_code_prop`, in which the package name is detokenized. This detokenization makes the normal patching utilities using `\scantokens` unusable. This is not a big problem since it is rarely needed.

**Comment by muzimuzhi** [comment]
> 更新：[latex3/latex2e#385](https://github.com/latex3/latex2e/issues/385) 修复了当前 issue，已随 latex2e [dev-2021-05-01-1](https://github.com/latex3/latex2e/releases/tag/dev-2021-05-01-1) 发布。

问题已解决，关闭 issue
